chatsky/llm/wrapper.py

```diff
@@ -0,0 +1,182 @@
+        elif set(message_schema.keys()).intersection(set(Message.model_fields.keys())) != set():
+            # Case if the message_schema desribes Message structure
+            raise NotImplementedError
+        else:
+            # Case if the message_schema desribes Message.text structure
+            structured_model = self.model.with_structured_output(message_schema)
+            result = self.parser.invoke(structured_model.invoke(history))
+            result = Message(text=str(result))
+        
+        if result.annotations:
+            result.annotations["__generated_by_model__"] = self.name
+        else:
+            result.annotations = {"__generated_by_model__": self.name}
+        return result
+    
+    def condition(self, prompt: str, method: BaseMethod):
+        def process_input(ctx: Context, _: Pipeline) -> bool:
+            result = method(ctx, self.parser.invoke(self.model.invoke([prompt+'\n'+ctx.last_request.text])))
+            return result
+        return process_input
+    
+    # Helper functions for DeepEval custom LLM usage
+    def generate(self, prompt: str):
+        return self.model.invoke(prompt).content
+    
+    async def a_generate(self, prompt: str):
+        return self.generate(prompt)
+    
+    def load_model(self):
+        return self.model
+    
+    def get_model_name(self):
+        return self.name
+
+
+def llm_response(
+        model_name,
+        prompt="",
+        history=5,
+        filter_func: Callable=lambda x: True
+    ):
+    """
+    Basic function for receiving LLM responses.
+    :param ctx: Context object. (Assigned automatically)
+    :param pipeline: Pipeline object. (Assigned automatically)
+    :param model_name: Name of the model from the `Pipeline.models` dictionary.
+    :param prompt: Prompt for the model.
+    :param history: Number of messages to keep in history.
+    :param filter_func: filter function to filter messages that will go the models context.
+    """
+    def wrapped(ctx: Context, pipeline: Pipeline):
+        model = pipeline.models[model_name]
+        if model.system_prompt == "":
+            history_messages = []
+        else:
+            history_messages = [SystemMessage(model.system_prompt)]
+        if history == 0 or len(ctx.responses)==0 or len(ctx.requests)==0:
+            return model.respond(history_messages+[message_to_langchain(Message(prompt + "\n" + ctx.last_request.text))])
+        else:
+            pairs = zip([ctx.requests[x] for x in range(len(ctx.requests))],
+                     [ctx.responses[x] for x in range(len(ctx.responses))])
+            for req, resp in filter(lambda x: filter_func(x), list(pairs)[-history:]):
+                history_messages.append(message_to_langchain(req))
+                history_messages.append(message_to_langchain(resp, human=False))
+            history_messages.append(message_to_langchain(ctx.last_request, prompt=prompt))
+            return model.respond(history_messages)
```

Comment:
Why is `prompt` added as a human message?

Reply:
Because it is common for user to append specific prompt for their message. Alternatively we can pass it as a SystemMessage and then follow it up with HumanMessage, but I think it may or may not clutter the dialogue history, but possibly can help to avoid unexpected behaviour that can be caused by having several different prompts in history

Reply:
Looking at usage examples in the first tutorial I'd imagine that `prompt` should be appended to `system_prompt`.

Also, I thought that one of the purposes of separating human and system messages is to prevent prompt injection from human side.

Reply:
Usage example show the prompt that contradicts the system prompt, but I think it should be reworked as the intended use case for passing prompts in the certain nodes is changing behaviour of the bot according to the flow (for example) we are in. As an example it could be "sell_flow" in the barista bot from example. And in the prompt we can specify for example missing menu positions and things we need to ask the customer. And maybe the last node in the flow would then utilize message_schema to compound users order in the json and send this json to an admin_user. But this brings up an idea of flow-wise prompts...

I agree with the prompt injection prevention via SystemMessage, it seems reasonable.